godot-core/src/builtin/array.rs

```diff
@@ -187,59 +274,176 @@ impl Array {
         assert!(step != 0);
         self.as_inner()
             .slice(to_i64(begin), to_i64(end), step.try_into().unwrap(), true)
+            .assume_type()
     }
 
-    /// Returns the value at the specified index as a `Variant`. To convert to a specific type, use
-    /// the available conversion methods on `Variant`, such as [`Variant::try_to`] or
-    /// [`Variant::to`].
+    /// Appends another array at the end of this array. Equivalent of `append_array` in GDScript.
+    pub fn extend_array(&mut self, other: TypedArray<T>) {
+        self.as_inner().append_array(other.assume_type());
+    }
+
+    /// Returns the runtime type info of this array.
+    fn type_info(&self) -> TypeInfo {
+        let variant_type = VariantType::from_sys(
+            self.as_inner().get_typed_builtin().try_into().unwrap());
+        let class_name = self.as_inner().get_typed_class_name();
+        TypeInfo { variant_type, class_name }
+    }
+
+    /// Checks that the inner array has the correct type set on it for storing elements of type `T`.
+    ///
+    /// # Panics
+    ///
+    /// If the inner type doesn't match `T` or no type is set at all.
+    fn check_type(&self) {
+        assert_eq!(self.type_info(), TypeInfo::new::<T>());
+    }
+
+    /// Sets the type of the inner array. Can only be called once, directly after creation.
+    fn set_typed(&mut self) {
+        debug_assert!(self.is_empty());
+        debug_assert!(!self.type_info().is_typed());
+        let type_info = TypeInfo::new::<T>();
+        if type_info.is_typed() {
+            let script = Variant::nil();
+            self.as_inner().set_typed(
+                type_info.variant_type.sys().into(), type_info.class_name, script);
+        }
+    }
+}
+
+impl<T: VariantMetadata + FromVariant> TypedArray<T> {
+    /// Returns an iterator over the `TypedArray` by reference. Instead of references to elements
```

Comment:
The "by reference" is a bit unclear, as the return type is a value. Maybe "referencing the array"?

Reply:
Done.